ci/owncloud/upload_to_owncloud.py

```diff
@@ -0,0 +1,153 @@
+    return {
+        "webdav_login": credentials[0],
+        "webdav_password": credentials[2],
+        "webdav_hostname": default_hostname + credentials[0]
+    }
+
+
+def get_path_components(dirname):
+    """
+    get_path_components(dirname)
+
+    Gets a list of all subpaths needed to build a directory, including itself.
+
+    Parameters
+    ----------
+    dirname : str
+        A given directory path
+    Returns
+    -------
+    A python list of path strings
+    For an input 'a/b/c' returns ['a/b/c', 'a/b', 'a']
+    """
+    path_list = [dirname, ]
+    tmp = path.split(dirname)
+    while tmp[0] != "" and not path.ismount(tmp[0]):
+        path_list.append(tmp[0])
+        tmp = path.split(tmp[0])
+    path_list.reverse()
+    return path_list
+
+
+def mkdir_recursively(webdav_client, remote_path):
+    """
+    mkdir_recursively(remote_path)
+
+    Recursively creates all directories needed to create the input path.
+
+    Parameters
+    ----------
+    remote_path: str
+    """
+    pathlist = get_path_components(remote_path)
+    for p in pathlist:
+        try:
+            webdav_client.mkdir(p)
+        except BaseException as e:
+            debug(e)
+
+
+def upload(webdav_client, is_direct, remote_path, local_path, unique_id=None):
+    """
+    upload(webdav_client, is_direct, remote_path, local_path, unique_id)
+
+    Uploads the content of a directory to a directory in the remote webdav server
+
+    The remote directory's tail is a unique directory name.
+
+    Parameters
+    ----------
+
+    webdav_client : webdav3.client.Client
+        the remote dav server's client object
+    is_direct : bool
+        flag dictating whether the directory is uploaded directly into the remote path
+        or copied into a temporary folder first, and then into the remote path.
+
+    remote_path : str
+        path in the webdav server to be created and for the files to be copied to
+
+    local_path : str
+        local path to be copied from
+
+    unique_id : str, optional
+        a uuid to act as the identifier of the uploaded files in the remote directory
+    """
+    if not unique_id:
+        unique_id = str(uuid4())
+    message("""Uploading files from local directory {0} to remote directory {1},\n \
+         username is {2}, remote directory id is: {3}""".format(
+        local_path, remote_path, webdav_client.webdav.options["login"], webdav_client, unique_id))
+    remote_full_path = path.join(remote_path, unique_id)
+    mkdir_recursively(remote_full_path)
+    if not path.isdir(local_path):
+        raise FileNotFoundError("Local path location is not a directory")
+    if is_direct:
+        webdav_client.upload_sync(remote_full_path, local_path)
+    else:
+        remote_temp_path = path.join("temp_uploads", unique_id)
+        mkdir_recursively(remote_temp_path)
+        webdav_client.upload_sync(remote_temp_path, local_path)
+        webdav_client.copy(remote_temp_path, remote_full_path)
```

Comment:
Shouldn't this be move rather than copy?

Also, if the remote_full_path already exists, isn't it going to add a new subdirectory under remote_full_path, rather than copying/moving all the individual files and directories?